certbot/plugins/dns_cloudflare.py

```diff
@@ -0,0 +1,323 @@
+"""DNS Authenticator for Cloudflare."""
+import logging
+
+import zope.interface
+
+from time import sleep
+
+from acme import challenges
+
+from certbot import errors
+from certbot import interfaces
+
+from certbot.display import util as display_util
+
+from certbot.plugins import common
+
+import CloudFlare
+
+logger = logging.getLogger(__name__)
+
+
+@zope.interface.implementer(interfaces.IAuthenticator)
+@zope.interface.provider(interfaces.IPluginFactory)
+class Authenticator(common.Plugin):
+    """DNS  Authenticator for Cloudflare
+
+    This Authenticator uses the Cloudflare API to fulfill a dns-01 challenge.
+    """
+
+    description = 'Obtain certs using a DNS TXT record (if you are using Cloudflare for DNS).'
+
+    propagation_seconds = 10
+    _attempt_cleanup = False
+
+    def __init__(self, *args, **kwargs):
+        super(Authenticator, self).__init__(*args, **kwargs)
+
+        self.email = None
+        self.api_key = None
+
+    @classmethod
+    def add_parser_arguments(cls, add):
+        add('email',
+            help='Email address associated with Cloudflare account.')
+        add('api-key',
+            help='API key for Cloudflare account. ' +
+                 '(Which can be obtained from https://www.cloudflare.com/a/account/my-account)')
+
+    def prepare(self): # pylint: disable=missing-docstring
+        pass
+
+    def more_info(self): # pylint: disable=missing-docstring,no-self-use
+        return 'This plugin configures a DNS TXT record to respond to a dns-01 challenge using ' + \
+               'the Cloudflare API.'
+
+    def get_chall_pref(self, unused_domain): # pylint: disable=missing-docstring,no-self-use
+        return [challenges.DNS01]
+
+    def perform(self, achalls): # pylint: disable=missing-docstring
+        self._setup_credentials()
+
+        self._attempt_cleanup = True
+
+        responses = []
+        for achall in achalls:
+            self._perform_achall(achall)
+            responses.append(achall.response(achall.account_key))
+
+        # DNS updates take time to propagate and checking to see if the update has occurred is not
+        # reliable (the machine this code is running on might be able to see an update before
+        # the ACME server). So: we sleep for a short amount of time we believe to be long enough.
+        sleep(self.propagation_seconds)
+
+        return responses
+
+    def _perform_achall(self, achall):
+        """
+        Performs a dns-01 challenge by creating a DNS TXT record.
+
+        :param `~certbot.achallenges.AnnotatedChallenge` achall: the challenge to perform
+        :raises errors.PluginError: If the challenge cannot be performed
+        """
+
+        domain = achall.domain
+        record_name = achall.validation_domain_name(domain)
+        record_content = achall.validation(achall.account_key)
+        ttl = 120
```

Comment:
Apparently, 120 is an undocumented minimum; a lower TTL is what leads to the `1004 DNS Validation Error`.